File: cpp/src/ConsoleUtils.cpp
#include "ConsoleUtils.h"
#include <iostream>
#include <algorithm>
#include <cctype>
#include <limits>
#include <iomanip>
#include <sstream>

// ANSI color codes
const std::string ConsoleUtils::RESET = "\033[0m";
const std::string ConsoleUtils::RED = "\033[31m";
const std::string ConsoleUtils::GREEN = "\033[32m";
const std::string ConsoleUtils::YELLOW = "\033[33m";
const std::string ConsoleUtils::BLUE = "\033[34m";
const std::string ConsoleUtils::CYAN = "\033[36m";
const std::string ConsoleUtils::WHITE = "\033[37m";
const std::string ConsoleUtils::BOLD = "\033[1m";

void ConsoleUtils::printColored(const std::string& text, const std::string& color) {
    std::cout << color << text << RESET;
    std::cout.flush();
}

void ConsoleUtils::printlnColored(const std::string& text, const std::string& color) {
    std::cout << color << text << RESET << std::endl;
}

void ConsoleUtils::printSuccess(const std::string& message) {
    printlnColored("✅ " + message, GREEN);
}

void ConsoleUtils::printError(const std::string& message) {
    printlnColored("❌ " + message, RED);
}

void ConsoleUtils::printInfo(const std::string& message) {
    printlnColored(message, BLUE);
}

void ConsoleUtils::printWarning(const std::string& message) {
    printlnColored("⚠️ " + message, YELLOW);
}

void ConsoleUtils::printHeader() {
    printlnColored("╔══════════════════════════════════════════╗", CYAN);
    printlnColored("║          Video to Audio Converter       ║", CYAN);
    printlnColored("║                 (C++)                    ║", CYAN);
    printlnColored("╚══════════════════════════════════════════╝", CYAN);
    std::cout << std::endl;
}

bool ConsoleUtils::askConfirmation(const std::string& question, bool defaultValue) {
    std::string defaultText = defaultValue ? "[Y/n]" : "[y/N]";
    printColored(question + " " + defaultText + ": ", YELLOW);
    
    std::string input;
    std::getline(std::cin, input);
    input = trim(toLower(input));
    
    if (input.empty()) {
        return defaultValue;
    }
    
    return input == "y" || input == "yes";
}

std::string ConsoleUtils::askInput(const std::string& question) {
    printColored(question + ": ", YELLOW);
    std::string input;
    std::getline(std::cin, input);
    return trim(input);
}

int ConsoleUtils::selectFromMenu(const std::string& title, const std::vector<std::string>& options) {
    std::cout << std::endl;
    printlnColored(title, CYAN);
    printlnColored(std::string(title.length(), '='), CYAN);
    
    for (size_t i = 0; i < options.size(); ++i) {
        printColored(std::to_string(i + 1) + ". ", GREEN + BOLD);
        std::cout << options[i] << std::endl;
    }
    
    while (true) {
        printColored("\nSelect an option (1-" + std::to_string(options.size()) + "): ", YELLOW);
        
        std::string input;
        std::getline(std::cin, input);
        input = trim(input);
        
        try {
            int choice = std::stoi(input);
            if (choice >= 1 && choice <= static_cast<int>(options.size())) {
                return choice - 1; // Return zero-based index
            } else {
                printError("Please enter a number between 1 and " + std::to_string(options.size()));
            }
        } catch (const std::exception&) {
            printError("Please enter a valid number");
        }
    }
}

void ConsoleUtils::printProgress(const std::string& message) {
    printColored("🔄 " + message, BLUE);
    std::cout.flush();
}

void ConsoleUtils::clearLine() {
    std::cout << "\r" << std::string(80, ' ') << "\r";
    std::cout.flush();
}

void ConsoleUtils::printFileSize(const std::string& label, double sizeMB) {
    std::ostringstream oss;
    oss << "📊 " << label << ": " << std::fixed << std::setprecision(2) << sizeMB << " MB";
    printlnColored(oss.str(), YELLOW);
}

std::string ConsoleUtils::trim(const std::string& str) {
    size_t first = str.find_first_not_of(" \t\r\n");
    if (std::string::npos == first) {
        return "";
    }
    size_t last = str.find_last_not_of(" \t\r\n");
    return str.substr(first, (last - first + 1));
}

std::string ConsoleUtils::toLower(const std::string& str) {
    std::string result = str;
    std::transform(result.begin(), result.end(), result.begin(), ::tolower);
    return result;
}

File: java/src/main/java/com/hackyourfuture/videoconverter/ConsoleUtils.java
package com.hackyourfuture.videoconverter;

import java.util.List;
import java.util.Scanner;

/**
 * Utility class for enhanced console output and user interaction.
 * Pure Java implementation without external dependencies.
 */
public class ConsoleUtils {
    
    private static final Scanner scanner = new Scanner(System.in);
    
    // ANSI color codes
    public static final String RESET = "\u001B[0m";
    public static final String RED = "\u001B[31m";
    public static final String GREEN = "\u001B[32m";
    public static final String YELLOW = "\u001B[33m";
    public static final String BLUE = "\u001B[34m";
    public static final String CYAN = "\u001B[36m";
    public static final String WHITE = "\u001B[37m";
    public static final String BOLD = "\u001B[1m";
    
    /**
     * Print colored text to console
     */
    public static void printColored(String text, String color) {
        System.out.print(color + text + RESET);
    }
    
    /**
     * Print colored line to console
     */
    public static void printlnColored(String text, String color) {
        System.out.println(color + text + RESET);
    }
    
    /**
     * Print success message (green)
     */
    public static void printSuccess(String message) {
        printlnColored("✅ " + message, GREEN);
    }
    
    /**
     * Print error message (red)
     */
    public static void printError(String message) {
        printlnColored("❌ " + message, RED);
    }
    
    /**
     * Print info message (blue)
     */
    public static void printInfo(String message) {
        printlnColored(message, BLUE);
    }
    
    /**
     * Print warning message (yellow)
     */
    public static void printWarning(String message) {
        printlnColored("⚠️ " + message, YELLOW);
    }
    
    /**
     * Print a simple ASCII header
     */
    public static void printHeader() {
        printlnColored("╔══════════════════════════════════════════╗", CYAN);
        printlnColored("║          Video to Audio Converter       ║", CYAN);
        printlnColored("║                 (Java)                   ║", CYAN);
        printlnColored("╚══════════════════════════════════════════╝", CYAN);
        System.out.println();
    }
    
    /**
     * Ask user for yes/no confirmation
     */
    public static boolean askConfirmation(String question, boolean defaultValue) {
        String defaultText = defaultValue ? "[Y/n]" : "[y/N]";
        printColored(question + " " + defaultText + ": ", YELLOW);
        
        String input = scanner.nextLine().trim().toLowerCase();
        
        if (input.isEmpty()) {
            return defaultValue;
        }
        
        return input.equals("y") || input.equals("yes");
    }
    
    /**
     * Ask user for text input
     */
    public static String askInput(String question) {
        printColored(question + ": ", YELLOW);
        return scanner.nextLine().trim();
    }
    
    /**
     * Display a numbered selection menu and get user choice
     */
    public static int selectFromMenu(String title, List<String> options) {
        System.out.println();
        printlnColored(title, CYAN);
        printlnColored("═".repeat(title.length()), CYAN);
        
        for (int i = 0; i < options.size(); i++) {
            System.out.printf("%s%s%d%s. %s%n", 
                GREEN + BOLD, 
                "", i + 1,
                RESET,
                options.get(i));
        }
        
        while (true) {
            printColored("\nSelect an option (1-" + options.size() + "): ", YELLOW);
            try {
                String input = scanner.nextLine().trim();
                int choice = Integer.parseInt(input);
                
                if (choice >= 1 && choice <= options.size()) {
                    return choice - 1; // Return zero-based index
                } else {
                    printError("Please enter a number between 1 and " + options.size());
                }
            } catch (NumberFormatException e) {
                printError("Please enter a valid number");
            }
        }
    }
    
    /**
     * Print a progress indicator
     */
    public static void printProgress(String message) {
        printColored("🔄 " + message, BLUE);
        System.out.flush();
    }
    
    /**
     * Clear the current line (for updating progress)
     */
    public static void clearLine() {
        System.out.print("\r" + " ".repeat(80) + "\r");
        System.out.flush();
    }
    
    /**
     * Print file size information
     */
    public static void printFileSize(String label, double sizeMB) {
        printlnColored(String.format("📊 %s: %.2f MB", label, sizeMB), YELLOW);
    }
    
    /**
     * Cleanup resources (no-op for pure Java version)
     */
    public static void cleanup() {
        // No cleanup needed for pure Java version
    }
}
